File: bootstrap-neoforge/src/main/java/ru/hollowhorizon/hollowengine/neoforge/NeoForgeClientEvents.java
package ru.hollowhorizon.hollowengine.neoforge;

import com.mojang.blaze3d.vertex.VertexFormat;
import com.mojang.brigadier.CommandDispatcher;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.ShaderInstance;
import net.minecraft.commands.SharedSuggestionProvider;
import net.minecraft.resources.ResourceLocation;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.neoforge.client.event.*;
import net.neoforged.neoforge.common.NeoForge;
import net.neoforged.neoforge.event.entity.player.ItemTooltipEvent;
import ru.hollowhorizon.hollowengine.bootstrap.impl.BootstrapRuntimeManager;
import ru.hollowhorizon.hollowengine.bootstrap.runtime.EventBridge;
import ru.hollowhorizon.hollowengine.bootstrap.runtime.RuntimeBridge;

import java.io.IOException;
import java.util.function.Consumer;

public class NeoForgeClientEvents {
    private static final RuntimeBridge bridge = BootstrapRuntimeManager.bridge();
    private static final EventBridge events = bridge.events();

    public static void init(IEventBus modBus) {
        var forgeBus = NeoForge.EVENT_BUS;
        forgeBus.addListener(NeoForgeClientEvents::onRenderTooltips);
        forgeBus.addListener(NeoForgeClientEvents::onClientTick);
        forgeBus.addListener(NeoForgeClientEvents::registerClientCommands);
        forgeBus.addListener(NeoForgeClientEvents::onRenderOverlayPre);
        forgeBus.addListener(NeoForgeClientEvents::onRenderOverlayPost);
        forgeBus.addListener(NeoForgeClientEvents::onRenderHudPost);
        forgeBus.addListener(NeoForgeClientEvents::onCameraSetup);

        modBus.addListener(NeoForgeClientEvents::registerShaders);
        modBus.addListener(NeoForgeClientEvents::registerRenderers);
        modBus.addListener(NeoForgeClientEvents::registerKeyMappings);
        modBus.addListener(NeoForgeClientEvents::registerReloadListeners);

    }

    private static void registerReloadListeners(RegisterClientReloadListenersEvent event) {
        events.onRegisterClientReloadListeners(event::registerReloadListener);
    }

    @SuppressWarnings("unchecked")
    private static void registerClientCommands(RegisterClientCommandsEvent event) {
        events.onClientCommandRegistration((CommandDispatcher<SharedSuggestionProvider>) (Object) event.getDispatcher(), event.getBuildContext());
    }

    private static void registerShaders(RegisterShadersEvent event) {
        events.onRegisterShaders(new NeoForgeShaders(event));
    }

    private static void registerRenderers(EntityRenderersEvent.RegisterRenderers event) {
        events.onRegisterEntityRenderers(event::registerEntityRenderer);
        events.onRegisterBlockEntityRenderers(event::registerBlockEntityRenderer);
    }

    private static void registerKeyMappings(RegisterKeyMappingsEvent event) {
        events.onRegisterKeybindings(event::register);
    }

    private static void onRenderTooltips(ItemTooltipEvent event) {
        events.onGetTooltip(event.getItemStack(), event.getContext(), event.getFlags(), event.getToolTip());
    }

    private static void onClientTick(ClientTickEvent.Post event) {
        events.onClientTick(Minecraft.getInstance());
    }

    private static void onRenderOverlayPre(RenderGuiLayerEvent.Pre event) {
        // Forward every named layer by its id; the runtime decides which ones it cares about. This
        // covers all vanilla layers and any modded layer without a per-layer mapping here.
        if (bridge.onRenderOverlayPre(Minecraft.getInstance().getWindow(), event.getGuiGraphics(), event.getPartialTick().getGameTimeDeltaPartialTick(false), event.getName().toString())) {
            event.setCanceled(true);
        }
    }

    private static void onRenderOverlayPost(RenderGuiLayerEvent.Post event) {
        bridge.onRenderOverlayPost(Minecraft.getInstance().getWindow(), event.getGuiGraphics(), event.getPartialTick().getGameTimeDeltaPartialTick(false), event.getName().toString());
    }

    private static void onRenderHudPost(RenderGuiEvent.Post event) {
        bridge.onRenderHudPost(Minecraft.getInstance().getWindow(), event.getGuiGraphics(), event.getPartialTick().getGameTimeDeltaPartialTick(false));
    }

    private static void onCameraSetup(ViewportEvent.ComputeCameraAngles event) {
        var setup = bridge.onCameraSetup(event.getRenderer(), event.getCamera(), event.getYaw(), event.getPitch(), event.getRoll(), (float) event.getPartialTick());
        event.setYaw(setup.yaw());
        event.setPitch(setup.pitch());
        event.setRoll(setup.roll());
    }

    private record NeoForgeShaders(RegisterShadersEvent event) implements EventBridge.ShaderRegistration {
        @Override
        public void register(ResourceLocation id, VertexFormat vertexFormat, Consumer<ShaderInstance> loadCallback) throws IOException {
            var shader = new ShaderInstance(event.getResourceProvider(), id, vertexFormat);
            event.registerShader(shader, loadCallback);
        }
    }
}

File: runtime/src/main/java/ru/hollowhorizon/hollowengine/neoforge/NeoForgeClientEvents.kt
package ru.hollowhorizon.hollowengine.neoforge

//? if neoforge {

/*import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.ShaderInstance
import net.minecraft.world.item.Item
import net.neoforged.bus.api.IEventBus
import net.neoforged.neoforge.client.event.ClientTickEvent
import net.neoforged.neoforge.client.event.CustomizeGuiOverlayEvent
import net.neoforged.neoforge.client.event.EntityRenderersEvent
import net.neoforged.neoforge.client.event.RegisterClientReloadListenersEvent
import net.neoforged.neoforge.client.event.RegisterKeyMappingsEvent
import net.neoforged.neoforge.common.NeoForge
import ru.hollowhorizon.hollowengine.common.events.EventBus
import ru.hollowhorizon.hollowengine.common.events.post
import ru.hollowhorizon.hollowengine.common.events.registry.RegisterBlockEntityRenderersEvent
import ru.hollowhorizon.hollowengine.common.events.registry.RegisterEntityRenderersEvent
import ru.hollowhorizon.hollowengine.common.events.registry.RegisterKeyBindingsEvent
import ru.hollowhorizon.hollowengine.common.events.registry.RegisterReloadListenersEvent
import ru.hollowhorizon.hollowengine.common.events.registry.RegisterShadersEvent


class NeoForgeClientEvents(val modBus: IEventBus) {


        init {
            modBus.addListener(::registerShaders)
            modBus.addListener(::onRegisterKeys)
            modBus.addListener(::onEntityRenderers)
            modBus.addListener(::registerReloadListeners)
            NeoForge.EVENT_BUS.addListener(::onClientTick)
            NeoForge.EVENT_BUS.addListener(::onRenderTooltips)
            NeoForge.EVENT_BUS.addListener(::onCameraSetup)
        }

        private fun onEntityRenderers(event: EntityRenderersEvent.RegisterRenderers) {
            RegisterEntityRenderersEvent(event::registerEntityRenderer).post()
            RegisterBlockEntityRenderersEvent(event::registerBlockEntityRenderer).post()
        }

        private fun registerReloadListeners(event: RegisterClientReloadListenersEvent) {
            val hcevent = RegisterReloadListenersEvent.Client()
            EventBus.post(hcevent)
            hcevent.listeners.forEach {
                event.registerReloadListener(it)
            }
        }

        private fun registerShaders(event: net.neoforged.neoforge.client.event.RegisterShadersEvent) {
            val hcEvent = RegisterShadersEvent()
            EventBus.post(hcEvent)
            hcEvent.shaders.forEach {
                event.registerShader(ShaderInstance(event.resourceProvider, it.key, it.value.first), it.value.second)
            }
        }

        private fun onClientTick(event: ClientTickEvent.Post) {
            EventBus.post(
                ru.hollowhorizon.hollowengine.common.events.tick.TickEvent.Client(
                    Minecraft.getInstance()
                )
            )
        }

        private fun onRegisterKeys(event: RegisterKeyMappingsEvent) {
            RegisterKeyBindingsEvent(event::register).post()
        }

        private fun onRenderTooltips(event: net.neoforged.neoforge.event.entity.player.ItemTooltipEvent) {
            ru.hollowhorizon.hollowengine.common.events.client.ItemTooltipEvent(
                event.flags,
                event.itemStack,
                event.toolTip,
                //? if >=1.21 {
                /^Item.TooltipContext.of(Minecraft.getInstance().level)
                ^///?}
            ).post()
        }

        private fun onCameraSetup(event: net.neoforged.neoforge.client.event.ViewportEvent.ComputeCameraAngles) {
            val hcEvent = ru.hollowhorizon.hollowengine.common.events.client.CameraSetupEvent(event.renderer, event.camera, event.partialTick.toFloat(), event.yaw, event.pitch, event.roll)
            hcEvent.post()
            event.yaw = hcEvent.yaw
            event.pitch = hcEvent.pitch
            event.roll = hcEvent.roll
        }
    
}

*///?}